### ISSUE TITLE
gunicorn worker is hanging on "Worker exiting"

### ISSUE THREAD

**Issue by talboren** [benoitc/gunicorn#2706]
I have flask server running and handling requests, our setup currently allows only 1 concurrent request (we're running in GCP cloud run where only 1 concurrent request is allowed) and therefor our gunicorn configuration is the following:

`gunicorn --preload --env APP_HOME=$APP_HOME --timeout 30 --bind :$PORT --workers 2 --max-requests 1 --log-level debug main:gunicorn_app`

we also require every worker to restart between requests and that's the reason we have --max-requests set to 1.
workers is set to 2 so whenever the first worker finishes his work, and restarts, another worker can already handle the next upcoming request.

What happens and how did I catch it?
A first requests gets to the service and a response is returned after 5~ seconds and the following log is logged:
> Worker exiting (pid: nnnnn)

A second requests gets to the second (booted) worker and a response is returned after another 5~ seconds, the following logs is again logged:
> Worker exiting (pid: nnnnn)

after 25~ seconds (please notice the 30 seconds timeout setting ^) the worker receives (and a log is logged) 
> WORKER TIMEOUT (pid:nnnnn) 

and an exception is thrown:
>Exception ignored in: <module 'threading' from '/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py'>
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/threading.py", line 1307, in _shutdown
    lock.acquire()
  File "/Users/talboren/Documents/Repositories/connectors/.venv/lib/python3.7/site-packages/gunicorn/workers/base.py", line 203, in handle_abort
    sys.exit(1)
SystemExit: 1

only after that exception is thrown, the worker seems to be booting again (and another log is logged):
> Booting worker with pid: nnnnn

I was able to reproduce the issue with a simple endpoint that basically just does the following:

1. creates a thread
2. thread sleeps for 31 seconds
3. return an empty response

And what this actually causes is that when the 2 workers are "stale" (they're exiting but they got stuck after trying to exit because of some dangling thread), GCP cloud run acknowledges the service as "ready to receive another request" while no worker is available, GCP (which also has 30 seconds configured for request timeout) then throws a "504 Timeout" error when the underlying service failed to answer within 30 seconds.

Hope I made the issue clear, I can of course try to share a demo application (which basically does what I've described above) but the complete setup has to be ran on GCP cloud run.

Gunicorn configuration:
> Current configuration:
  config: ./gunicorn.conf.py
  wsgi_app: None
  bind: [':8080']
  backlog: 2048
  workers: 2
  worker_class: sync
  threads: 1
  worker_connections: 1000
  max_requests: 1
  max_requests_jitter: 0
  timeout: 30
  graceful_timeout: 30
  keepalive: 2
  limit_request_line: 4094
  limit_request_fields: 100
  limit_request_field_size: 8190
  reload: False
  reload_engine: auto
  reload_extra_files: []
  spew: False
  check_config: False
  print_config: False
  preload_app: True
  sendfile: None
  reuse_port: False
  chdir: /Users/talboren/Documents/Repositories/connectors
  daemon: False
  raw_env: ['APP_HOME=/Users/talboren/Documents/Repositories/connectors']
  pidfile: None
  worker_tmp_dir: None
  user: 501
  group: 20
  umask: 0
  initgroups: False
  tmp_upload_dir: None
  secure_scheme_headers: {'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}
  forwarded_allow_ips: ['127.0.0.1']
  accesslog: None
  disable_redirect_access_to_syslog: False
  access_log_format: %(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s"
  errorlog: -
  loglevel: debug
  capture_output: False
  logger_class: gunicorn.glogging.Logger
  logconfig: None
  logconfig_dict: {}
  syslog_addr: unix:///var/run/syslog
  syslog: False
  syslog_prefix: None
  syslog_facility: user
  enable_stdio_inheritance: False
  statsd_host: None
  dogstatsd_tags: 
  statsd_prefix: 
  proc_name: None
  default_proc_name: main:gunicorn_app
  pythonpath: None
  paste: None
  on_starting: <function OnStarting.on_starting at 0x7f809d27e440>
  on_reload: <function OnReload.on_reload at 0x7f809d27e560>
  when_ready: <function WhenReady.when_ready at 0x7f809d27e680>
  pre_fork: <function Prefork.pre_fork at 0x7f809d27e7a0>
  post_fork: <function Postfork.post_fork at 0x7f809d27e8c0>
  post_worker_init: <function PostWorkerInit.post_worker_init at 0x7f809d27e9e0>
  worker_int: <function WorkerInt.worker_int at 0x7f809d27eb00>
  worker_abort: <function WorkerAbort.worker_abort at 0x7f809d27ec20>
  pre_exec: <function PreExec.pre_exec at 0x7f809d27ed40>
  pre_request: <function PreRequest.pre_request at 0x7f809d27ee60>
  post_request: <function PostRequest.post_request at 0x7f809d27eef0>
  child_exit: <function ChildExit.child_exit at 0x7f809d296050>
  worker_exit: <function WorkerExit.worker_exit at 0x7f809d296170>
  nworkers_changed: <function NumWorkersChanged.nworkers_changed at 0x7f809d296290>
  on_exit: <function OnExit.on_exit at 0x7f809d2963b0>
  proxy_protocol: False
  proxy_allow_ips: ['127.0.0.1']
  keyfile: None
  certfile: None
  ssl_version: 2
  cert_reqs: 0
  ca_certs: None
  suppress_ragged_eofs: True
  do_handshake_on_connect: False
  ciphers: None
  raw_paste_global_conf: []
  strip_header_spaces: False

* It is important to note that we cannot control underlying threads because we're using many third-party libraries that may have threads started and dangling that are out of our control, so I am pretty lost in here, when running the above mentioned steps while setting the thread to be a daemon, it seems like the worker exits just fine.

Is there a way that I can make max_requests force kill the worker and not just a "soft" kill that fails due to the running thread?

**Comment by hynek** [comment]
I don't know why, but this (not exactly same setup but very similar with workers=1 and extremely low load) has become a daily occurrence with _some_ apps for us. Moving them to uWSGI and it's…less delicate way of killing processes…fixed it.

**Comment by benoitc** [comment]
@talboren i just noticed this ticket. So far going back from some retreat. The info you posted doesn't show something unexpected: https://docs.gunicorn.org/en/stable/design.html 

I don't know what does your application but it seems to block the worker sttopping it to infor the arbiter it's alive. I know it was a long time ago but if you can give more info about what were doing your application that would be helpful.

**Comment by hynek** [comment]
> @hynek how does it helps? How to reproduce the issues you got? Which kind of app?

Unfortunately it takes like a day to manifest and hasn't with uWSGI yet. It happens to me with a falcon-based web app that has an extra process and with our httpbin instance. They just freeze, but can be unfrozen by sending a HUP. Unfortunately the environment doesn't allow me to strace.

**Comment by benoitc** [comment]
no activity since awhile. closing feel free to create a new ticket if needed.
